### PR TITLE
Add wrapper module for cryptographic primitives

### LINKED ISSUE #29: Implement Network.Ricochet.Crypto
We will need a module with nice crypto-related utility functions.
One thing we need is the generation of 1024-Bit private RSA keys in base64 for `network-anonymous-tor`.

### PR DESCRIPTION
**Pull request by photm5**
Fixes #29.  I think this is everything we need.  Do you see something I missed?

### CODE REVIEW

**Comment by photm5** [comment]
@froozen I fixed the shell.nix problem you introduced in #36, does it still work for you?

**EDIT:** Oops, still not fixed...  Did it really work for you?

**Comment by froozen** [comment]
Yeah, it worked perfectly fine.

**Comment by photm5** [comment]
Did you try forcing cabal to reconfigure using `cabal configure`? This is where the problem happens, since it seems to forget the new `hprotoc` version. (It errors since the old version doesn’t know the `--lenses` flag.)

**Comment by froozen** [comment]
Both `cabal configure` and `cabal build` run fine for me.

**Comment by photm5** [comment]
Weird... Could you try if it still works at ca34250385ea2b81ad9803a7d9cb35c8c33c6e30?

**Comment by photm5** [comment]
Oh, wait a minute, I changed the network-anonymous-tor version by accident...

**Comment by froozen** [comment]
`shell.nix` seems to be messed up in general.

**Comment by photm5** [comment]
Could you try out fa93becaac1a2f80e99ce795c26e96b77bd2c031 please?

File: src/Network/Ricochet/Crypto/RSA.hsc [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+-- | Parse a private key from ASN.1 DER format
+fromDERPriv :: RSAKey k
+            => ByteString -- ^ The private key information encoded in ASN.1 DER
+            -> Maybe k    -- ^ This can return either 'RSAPubKey' or
+                          --   'RSAKeyPair' because there’s sufficient
+                          --   information for both.
+fromDERPriv = makeDecodeFun _fromDERPriv
+
+-- vim: ft=haskell
```
Comment (froozen) [inline src/Network/Ricochet/Crypto/RSA.hsc]: Please remove  this to keep files editor agnostic.

**Comment by photm5** [comment]
It doesn’t help that GitHub spits out the whole commit log again whenever I change one single commit in the history.

File: src/Network/Ricochet/Crypto/RSA.hsc (hunk 1 above)
Comment (photm5) [inline src/Network/Ricochet/Crypto/RSA.hsc]: Ok :)... Was going to add a line note asking a question about it myself, but got distracted.

**Comment by froozen** [comment]
`fa93be` works fine for me. The expression looks disguting, though.

**Comment by froozen** [comment]
What is your nix-version? I'm on 1.10.

**Comment by photm5** [comment]
It does.  When I first added this dirt to the codebase, I tried to refactor it myself (Using foldl, too), but failed horribly.  You seem to be a bit more proficient with Nix, as you got it to work with your setup at least :)

**Comment by photm5** [comment]
I’m on 1.9.  However, I think the problem is not caused by our Nix versions, but rather the NixOS channel (I didn’t upgrade for a while), since Nix itself tries to stay a lot more compatible than the nixpkgs are.

**Comment by froozen** [comment]
Could you try upgrading and see if it works, then?

**Comment by photm5** [comment]
Yes, it does work with the current Nix and Nix expressions.  Finally removing that ugly `Fix shell.nix` commit.

**Comment by froozen** [comment]
Great. Yay for higher-order abstractions!

File: src/Network/Ricochet/Crypto/RSA.hsc [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+makeDecodeFun :: RSAKey k => CDecodeFun -> ByteString -> Maybe k
+makeDecodeFun fun bs = unsafePerformIO . usingConvedBS $ \(csPtr, ci) -> do
+  rsaPtr <- fun nullPtr csPtr ci
+  if rsaPtr == nullPtr then return Nothing else absorbRSAPtr rsaPtr
+  where usingConvedBS io = B.useAsCStringLen bs $ \(cs, len) ->
+          alloca $ \csPtr -> poke csPtr cs >> io (csPtr, fromIntegral len)
+
+makeEncodeFun :: RSAKey k => CEncodeFun -> k -> ByteString
+makeEncodeFun fun k = unsafePerformIO $ do
+  requiredSize <- withRSAPtr k $ flip _toDERPub nullPtr
```
Comment (froozen) [inline src/Network/Ricochet/Crypto/RSA.hsc]: Could you please explain what this line does? I can't figure it out.

**Comment by froozen** [comment]
Is there a plan to add lenses for those functions? It would make sense for a lot of them.

File: src/Network/Ricochet/Crypto.hs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+generate1024BitRSA :: IO RSAKeyPair
+generate1024BitRSA = generateRSAKey' 1024 3
+
+-- | Encode a private RSA key in DER and then Base64
+base64EncodePrivRSA :: RSAKeyPair -> ByteString
+base64EncodePrivRSA = encodeBase64BS . toDERPriv
+
+-- | The SHA256 Hash algorithm
+sha256 :: Digest
+sha256 = fromJust . unsafePerformIO . withOpenSSL $ getDigestByName "SHA256"
```
Comment (sternenseemann) [inline src/Network/Ricochet/Crypto.hs:46]: Uuuuuh, unsafePerformIO…

Comment (froozen) [inline src/Network/Ricochet/Crypto.hs:46]: That's standard for C-Bindings, as they are inside the IO monad by default.

Comment (sternenseemann) [inline src/Network/Ricochet/Crypto.hs:46]: I know, just the first time I see it actually being used :D

File: src/Network/Ricochet/Crypto/RSA.hsc (hunk 2 above)
Comment (sternenseemann) [inline src/Network/Ricochet/Crypto/RSA.hsc]: Maybe add some haddock comments plus comments explaining smaller fractions of the code?

**Comment by photm5** [comment]
Do you miss any primitive or lens?

**Comment by froozen** [comment]
Looks good to me, you can merge. :+1:

**Pull request merged**